### DISCUSSION TITLE
What is the best way to access state specific data?

### DISCUSSION THREAD

**Discussion by ma-hnln** [bevyengine/bevy#10327] [Q&A]
While using states I would like to access data which is specific to one state variant. In the below example this would be the `Entity` inside the `Active` variant. 

```rust
#[derive(Default, States, ...)]
enum SomeState {
    #[default]
    Inactive,
    Active(Entity),
}
```

I know that it is possible to query the state from any system like follows.

```rust
fn some_system(state: Res<State<SomeState>>);
```

I then always have to check, which state variant is actually active. Even inside systems which specifically run only within a single state variant (e.g. `Active`), I still have to do this. 

Can a system running only in one state variant directly access the state as if it was this exact variant?

**Comment by Jacob-JB** [comment]
The way I understand it, `Active(0v0)` and `Active(1v0)` would be considered two separate states not the same state with different data. The actual state value is meant to be small and is frequently cloned. You may be able to get away with this pattern with some manual PartialEq impls or other trickery but I don't believe this is the intended way of having state data.

The way I've always handled this is by making use of the OnEnter() and OnExit() schedules to manage resources that only exist when in certain states. Remember `Option<Res<T>>` is a valid system parameter if you need a system that does separate things based on state and the resource might not always exist.

Remember that bevy is designed for you to pick and choose the functionality that you want, if the state system has a problem that causes it to not work for your case, such as wanting the safety of *if my enum is this variant, it is garunteed to have this data*, just write your own state machine. Bevy's state is just built with the same tools you have available, so theres no inherent performance cost to replacing it with different logic.
